keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+- https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/util/flushwriter/writer.go#L50
+
+
+### Termination of the Request Handler
+
+Upon timeout, a request handler can finish its run by returning the error it encounters while:
+- reading from the `Body` of the Request
+- writing via the `Write` method of the `ResponseWriter` object, or
+- flushing using the `FlushError` method
+
+
+Another option is to wrap the `ResponseWriter`, and the `Body` of the request, and panic on timeout
+```
+type timeoutWriter struct {
+	http.ResponseWriter
+}
+
+func (tw *timeoutWriter) Write(p []byte) (int, error) {
+	n, err := tw.ResponseWriter.Write(p)
+	if errors.IsTimeout(err) {
+		panic(http.ErrAbortHandler)
+	}
+	return n, err
+}
+
+func (tw *timeoutWriter) FlushError() error {
+	err := tw.ResponseWriter.FlushError()
+	if errors.IsTimeout(err) {
+		panic(http.ErrAbortHandler)
+	}	
+}
+
+
+type timeoutReader struct {
+	io.ReadCloser
+}
+
+func (tr *timeoutReader) Read(p []byte) (n int, err error) {
+	n, err := tr.ReadCloser.Read(p)
+	if errors.IsTimeout(err) {
+		panic(http.ErrAbortHandler)
+	}	
+	return n, err	
+}
+
+
+// wrap inside the WithPerRequestTimeout filter
+func(w http.ResponseWriter, req *http.Request) {
+	tr := &timeoutReader{ReadCloser: req.Body}
+	req = req.Clone(req.Context())
+	req.Body = tr
+	w = &baseTimeoutWriter{ResponseWriter: w}
+	
+	handler.ServeHTTP(req, w)
+}
+```
+
+### Convert the Asynchronous Timeout Filter for Mutating-Request to Synchronous:
+
+In addition to the `timeout` filter, the admission and rest logic of mutating-requests are executed 
+asynchronously in their own goroutines by the [asynchronous finisher](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go#L87-L122): 
+
+- [create](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L193-L218)
+- [delete](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go#L128-L132)
+- [deletecollection](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go#L270-L272)
+- [patch](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L667-L689)
+- [update](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/update.go#L225-L237)
+
+We can replace the asynchronous finisher with one that executes the request
```

Comment:
the finisher is very old code [10 years?](https://github.com/kubernetes/kubernetes/blame/release-1.1/pkg/apiserver/resthandler.go#L618) does anyone remember why it's async?

Reply:
the only thing that i can think of is, it predated context maybe?